Automattic/wp-calypso
# Comments: Adapt QuerySiteComments to comments trees

Closes #16154

Currently `QuerySiteComments` requests comments via [`GET /sites/$site/comments/`](https://developer.wordpress.com/docs/api/1.1/get/sites/$site/comments/), which fetches a maximum of 100 entire comment objects.

## Proposed changes

- Replace `GET /sites/$site/comments/` with `GET /sites/$site/comments-tree/`.
This new endpoint fetches only the `commentId`s (and the IDs of the comment relationships, such as `postId` and `parentCommentId`, in order to eventually create a tree).

- The new `COMMENTS_TREE_REQUEST` and `COMMENTS_TREE_RECEIVE` action types handle the comments tree endpoint between Redux and the data layer.
On request success, the Redux state is updated to a structure like this:
```
comments: {
  trees: {
    [ siteId ]: {
      [ commentId ]: {
        commentId: 123,
        parentCommentId: 456,
        postId: 789,
      }
    }
  },
  counts: {
    [ siteId ]: {
      [ commentId ]: {
        approved: 1,
        spam: 2,
        trash: 3,
      }
    }
  }
}
```

- Whenever `comments.tree` changes, the query component fetches the new comment objects and updates the normal `comments.items` list.
In the current implementation, it dispatches a loop of `requestComment` actions. Eventually it will be paginated and/or lazy loaded.

## Testing instructions

- Open `/comments/approved/{siteSlug}` (possibly one with few comments).
- Check that the Redux `comments.trees.{siteId}.{status}` is populated with a commentId-keyed object of comments.
- Check that the Redux `comments.counts.{siteId}` is populated with a status-keyed object of counts.
- Check that a network request for each comment in `comments.trees.{siteId}.{status}` is dispatched.
- Check that the comments are then loaded and displayed.
- Navigate to another status list and repeat the checks, making sure that the previous status tree and count are not overwritten in the process.

## Notes

- The comment tree endpoint is currently only enabled for ~Simple sites and~ Automattic users.
- The endpoint does not currently support the "unapproved" status (see D6569-code).
